examples/hparam-tuning/tune.py

```diff
@@ -0,0 +1,199 @@
+
+            # zero the parameter gradients
+            optimizer.zero_grad()
+
+            # forward + backward + optimize
+            outputs = net(inputs)
+            loss = criterion(outputs, labels)
+            loss.backward()
+            optimizer.step()
+
+            # print statistics
+            running_loss += loss.item()
+            epoch_steps += 1
+            if i % 2000 == 1999:  # print every 2000 mini-batches
+                print(
+                    "[%d, %5d] loss: %.3f"
+                    % (epoch + 1, i + 1, running_loss / epoch_steps)
+                )
+                running_loss = 0.0
+
+        # Validation loss
+        val_loss = 0.0
+        val_steps = 0
+        total = 0
+        correct = 0
+        for i, data in enumerate(valloader, 0):
+            with torch.no_grad():
+                inputs, labels = data
+                inputs, labels = inputs.to(device), labels.to(device)
+
+                outputs = net(inputs)
+                _, predicted = torch.max(outputs.data, 1)
+                total += labels.size(0)
+                correct += (predicted == labels).sum().item()
+
+                loss = criterion(outputs, labels)
+                val_loss += loss.cpu().numpy()
+                val_steps += 1
+
+        # Here we save a checkpoint. It is automatically registered with
+        # Ray Tune and can be accessed through `session.get_checkpoint()`
+        # API in future iterations.
+        os.makedirs("my_model", exist_ok=True)
+        torch.save((net.state_dict(), optimizer.state_dict()), "my_model/checkpoint.pt")
+        checkpoint = Checkpoint.from_directory("my_model")
+
+        session.report(
+            {"loss": (val_loss / val_steps), "accuracy": correct / total},
+            checkpoint=checkpoint,
+        )
+    print("Finished Training")
+
+
+def main(num_samples=10, max_num_epochs=10, gpus_per_trial=2):
+    config = {
+        "l1": tune.choice([8, 16, 32, 64, 82, 120]),
+        "l2": tune.choice([8, 16, 32, 64, 82, 120]),
+        "lr": tune.loguniform(1e-4, 1e-1),
+        "batch_size": tune.choice([2, 4, 8, 16]),
+    }
+    scheduler = ASHAScheduler(max_t=max_num_epochs, grace_period=1, reduction_factor=2)
+
+    tuner = tune.Tuner(
+        tune.with_resources(
+            tune.with_parameters(train_cifar),
+            resources={"cpu": 4, "gpu": gpus_per_trial},
+        ),
+        tune_config=tune.TuneConfig(
+            metric="loss",
+            mode="min",
+            scheduler=scheduler,
+            num_samples=num_samples,
+        ),
+        run_config=air.RunConfig(
+            name="mlflow",
+            callbacks=[
+                MLflowLoggerCallback(
+                    tracking_uri="./mlruns",
+                    experiment_name="cifar10_tune",
+                )
+            ],
+        ),
+        param_space=config,
+    )
+
+    tuner.fit()
+
+
+main(num_samples=20, max_num_epochs=15, gpus_per_trial=1)
```

Comment:
@tmynn if I'm correct this requires that the system has GPU, maybe default this to 0, and add a command line argument to enable tuning with GPUs?